staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go

```diff
@@ -81,17 +81,35 @@ func (f *dynamicSharedInformerFactory) ForResource(gvr schema.GroupVersionResour
 	return informer
 }
 
-// Start initializes all requested informers.
-func (f *dynamicSharedInformerFactory) Start(stopCh <-chan struct{}) {
+func (f *dynamicSharedInformerFactory) start(stopCh <-chan struct{}) *sync.WaitGroup {
+	wg := &sync.WaitGroup{}
 	f.lock.Lock()
 	defer f.lock.Unlock()
 
 	for informerType, informer := range f.informers {
 		if !f.startedInformers[informerType] {
-			go informer.Informer().Run(stopCh)
+			wg.Add(1)
+			informerCopy := informer.Informer()
+			go func() {
+				defer wg.Done()
+				informerCopy.Run(stopCh)
+			}()
 			f.startedInformers[informerType] = true
 		}
 	}
+
+	return wg
+}
+
+// Start initializes all requested informers.
+func (f *dynamicSharedInformerFactory) Start(stopCh <-chan struct{}) {
+	f.start(stopCh)
+}
+
+// Start initializes all requested informers and blocks until they
+// all complete
```

Comment:
this actually feels a bit harder to utilize, at least in some cases. In some of the in-tree usages you can see you have to spin up a new goroutine just to wait which wasn't required with Shutdown ().

In our use case we actually call Start() quite a few times, because it's a shared factory across the entire application. So now we may end up with multiple goroutine all needing to wait.. and each wait is waiting for some possibly arbitrary subset of informers to exit. To be clear, I am thinking of the case where we start informer a, b, c then call run(), later start d, e and call run(). A-E are all different components of the code. Now I want to wait for all of them to shutdown

Reply:
Thinking about it more, if you have anywhere in the code call Start() and not Run() as well, you are at risk of missing some informers. 

Overall I am really preferring the Shutdown() approach (aside from the name, perhaps)...

Reply:
I'm still really confused about the semantics of this API then. You can call run multiple times, with different stopCh, what happens if you only close some of these and then call Shutdown? is the stopCh closing everything or only the one that had not started when you called Run? I just have no idea by looking at the type how this will work.

Reply:
Shutdown would block. As mentioned in the documentation:

> In addition, Shutdown blocks until all goroutines have terminated. For that
>  to happen, the close channel(s) that they were started with must be closed,
> either before Shutdown gets called or while it is waiting.

Reply:
So a valid sequence would be:
- Start(stopCh1)
- Start(stopCh2)
- close(stopCh1)
- close(stopCh2)
- Shutdown

Reply:
Is this a valid sequence:
```
Start(stopCh1)
close(stopCh1)
Start(stopCh2)
close(stopCh2)
Shutdown
```

Reply:
@apelisse: sorry, I missed you question. Yes, that's a valid sequence.

@ howardjohn: the "Shutdown" name came out of the discussion in https://github.com/kubernetes/kubernetes/pull/112200#discussion_r964306374 when it was suggested that in addition to waiting, calling that function should also prevent new goroutines from being started. In that sense it "shuts down" the factory.